rclcpp/include/rclcpp/publisher.hpp

```diff
@@ -230,97 +90,77 @@ class Publisher : public PublisherBase
   virtual void
   publish(std::unique_ptr<MessageT, MessageDeleter> & msg)
   {
-    bool inter_process_subscriptions_exist =
-      get_subscription_count() > get_intra_process_subscription_count();
-    if (!intra_process_is_enabled_ || inter_process_subscriptions_exist) {
+    if (!intra_process_is_enabled_) {
       this->do_inter_process_publish(msg.get());
-    }
-    if (store_intra_process_message_) {
-      // Take the pointer from the unique_msg, release it and pass as a void *
-      // to the ipm. The ipm should then capture it again as a unique_ptr of
-      // the correct type.
-      // TODO(wjwwood):
-      //   investigate how to transfer the custom deleter (if there is one)
-      //   from the incoming unique_ptr through to the ipm's unique_ptr.
-      //   See: http://stackoverflow.com/questions/11002641/dynamic-casting-for-unique-ptr
-      MessageT * msg_ptr = msg.get();
-      msg.release();
-      uint64_t message_seq =
-        store_intra_process_message_(intra_process_publisher_id_, msg_ptr, typeid(MessageT));
-      rcl_interfaces::msg::IntraProcessMessage ipm;
-      ipm.publisher_id = intra_process_publisher_id_;
-      ipm.message_sequence = message_seq;
-      auto status = rcl_publish(&intra_process_publisher_handle_, &ipm);
-      if (RCL_RET_PUBLISHER_INVALID == status) {
-        rcl_reset_error();  // next call will reset error message if not context
-        if (rcl_publisher_is_valid_except_context(&intra_process_publisher_handle_)) {
-          rcl_context_t * context = rcl_publisher_get_context(&intra_process_publisher_handle_);
-          if (nullptr != context && !rcl_context_is_valid(context)) {
-            // publisher is invalid due to context being shutdown
-            return;
-          }
-        }
-      }
-      if (RCL_RET_OK != status) {
-        rclcpp::exceptions::throw_from_rcl_error(status, "failed to publish intra process message");
-      }
-    } else {
-      // Always destroy the message, even if we don't consume it, for consistency.
       msg.reset();
+      return;
+    }
+    std::unique_ptr<MessageT, MessageDeleter> msg_copy;
+    // If interprocess subscriptors exists, then promote to shared and publish.
```

Comment:
`subscriptions`, or if you must `subscribers`. I don't think `subscriptors` is a word.

Reply:
Also the more documentation the better in cases like this. I'd devote a paragraph of comments here to describe that this is an optimization for a common, but narrow case and that it is important for performance. This will help people looking to understand the code for debugging or education and it makes it less likely that this code gets removed accidentally long down the road.